File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.cpp
class Solution {
public:
    string reverseWords(string s) 
    {
        // string res;
        // string part;
        // int n = s.length();
        // for(int i = 0; i < n; i++)
        // {


        //     if(s[i] == ' ')
        //     {
        //         reverse(part.begin(),part.end());
        //         res += part + ' ';
        //         part = "";
        //     }
        //     else
        //     {
        //         part += s[i];
        //     }

        //     if(i == (n - 1))
        //     {
        //         reverse(part.begin(),part.end());
        //         res += part;
        //     }
        // }

        // return res;

        int l = 0;
        int r = 0;
        for(int i = 0; i < s.length(); i++)
        {
            if(s[i] == ' '|| i == (s.length() - 1))
            {
                if(i == (s.length() - 1))
                {
                    r = i;
                }
                else
                {
                    r = i - 1;
                }
                
                while(l < r)
                {
                    char temp = s[l];
                    s[l] = s[r];
                    s[r] = temp;
                    l++;
                    r--;
                }
                if(i != (s.length() - 1))
                {
                    l = i + 1;
                }
                
            }
        }

        return s;
        
    }
};

File: 0557-reverse-words-in-a-string-iii/0557-reverse-words-in-a-string-iii.java
class Solution {

    public void reverse(char[] ch, int l, int r)
    {
        while(l < r)
        {
            char temp = ch[l];
            ch[l] = ch[r];
            ch[r] = temp;
            l++;
            r--;
        }
    }

    public String reverseWords(String s) {
        int n = s.length();
        char[] ch = s.toCharArray();
        int start = 0;

        for(int i = 0; i < n; i++)
        {
            if(ch[i] == ' ')
            {
                reverse(ch,start,i-1);
                start = i + 1;
            }
        }

        reverse(ch,start,n-1);
        StringBuilder sb = new StringBuilder("");
        return sb.append(ch).toString();
    }
}